packages/babel-types/src/traverse/traverse.js

```diff
@@ -1,65 +1,47 @@
-import { VISITOR_KEYS } from "./index";
-
-/**
- * A prefix AST traversal implementation meant for simple searching
- * and processing.
- */
-export function traverseFast(
-  node: Node,
-  enter: (node: Node) => void,
-  opts?: Object,
-) {
-  if (!node) return;
-
-  const keys = VISITOR_KEYS[node.type];
-  if (!keys) return;
-
-  opts = opts || {};
-  enter(node, opts);
-
-  for (const key of keys) {
-    const subNode = node[key];
-
-    if (Array.isArray(subNode)) {
-      for (const node of subNode) {
-        traverseFast(node, enter, opts);
-      }
-    } else {
-      traverseFast(subNode, enter, opts);
-    }
-  }
-}
+// @flow
+import { VISITOR_KEYS } from "../definitions";
 
 export type TraversalAncestors = Array<{
   node: BabelNode,
   key: string,
   index?: number,
 }>;
-export type TraversalHandler<T> = (BabelNode, TraversalAncestors, T) => void;
-export type TraversalHandlers<T> = {
-  enter?: TraversalHandler<T>,
-  exit?: TraversalHandler<T>,
+export type TraversalHandler<T: BabelNode, S: Object> = (
+  T,
+  TraversalAncestors,
+  S,
+) => void;
+export type TraversalHandlers<T: BabelNode, S: Object> = {
+  enter?: TraversalHandler<T, S>,
+  exit?: TraversalHandler<T, S>,
 };
 
 /**
  * A general AST traversal with both prefix and postfix handlers, and a
  * state object. Exposes ancestry data to each handler so that more complex
  * AST data can be taken into account.
  */
-export function traverse<T>(
-  node: BabelNode,
-  handlers: TraversalHandler<T> | TraversalHandlers<T>,
-  state?: T,
-) {
+export default function traverse<T: Object, S: Object>(
```

Comment:
This type signature also doesn't work. The type here implies that the type of the argument to the handler is the same type as the `node` argument, which is not the case. The handler is called with the node, and also any child node type, meaning it'll accept any `BabelNode` type.

Reply:
Yes sorry, I reverted all the type changes.